### PR TITLE
add param timeout for helm target provider

### PR DESCRIPTION
**Pull request by yanjiaxin534**

### CODE REVIEW

File: api/pkg/apis/v1alpha1/providers/target/helm/helm.go [hunk 1]
```diff
@@ -680,26 +689,52 @@ func pullOCIChart(ctx context.Context, repo, version string) (*registry.PullResu
...
 		installClient.Namespace = constants.DefaultScope
 	} else {
 		installClient.Namespace = deployment.Instance.Spec.Scope
 	}
+
 	installClient.Wait = componentProps.Wait
+	if componentProps.Timeout != "" {
+		duration, err := time.ParseDuration(componentProps.Timeout)
+		if err != nil {
+			sLog.Errorf("  P (Helm Target): failed to parse timeout duration: %v", err)
```
Comment (msftcoderdjw) [inline api/pkg/apis/v1alpha1/providers/target/helm/helm.go:705]: Change the method signature, pass context.Context as parameter and use sLog.ErrorfCtx(ctx, xxx) to print logs.

File: api/pkg/apis/v1alpha1/providers/target/helm/helm.go [hunk 2]
```diff
@@ -680,26 +689,52 @@ func pullOCIChart(ctx context.Context, repo, version string) (*registry.PullResu
...
+			return nil, err
+		}
+		if duration < 0 {
+			sLog.Errorf("  P (Helm Target): Timeout is negative: %s", componentProps.Timeout)
```
Comment (msftcoderdjw) [inline api/pkg/apis/v1alpha1/providers/target/helm/helm.go:709]: same as above.

File: api/pkg/apis/v1alpha1/providers/target/helm/helm.go [hunk 3]
```diff
@@ -710,12 +745,25 @@ func configureUpgradeClient(componentProps *HelmChartProperty, deployment *model
...
+	return upgradeClient, nil
 }
 
-func configureUninstallClient(deployment *model.DeploymentSpec, config *action.Configuration) *action.Uninstall {
+func configureUninstallClient(componentProps *HelmChartProperty, deployment *model.DeploymentSpec, config *action.Configuration) (*action.Uninstall, error) {
 	uninstallClient := action.NewUninstall(config)
-	return uninstallClient
+	uninstallClient.Wait = componentProps.Wait
+	if componentProps.Timeout != "" {
+		duration, err := time.ParseDuration(componentProps.Timeout)
```
Comment (msftcoderdjw) [inline api/pkg/apis/v1alpha1/providers/target/helm/helm.go:755]: Better to write a common function for parsing duration.

**Review by msftcoderdjw — CHANGES_REQUESTED** [review]

**Review by Haishi2016 — APPROVED** [review]

**Review by Haishi2016 — CHANGES_REQUESTED** [review]
Of course, approved only after it builds :)

File: api/pkg/apis/v1alpha1/providers/target/helm/helm.go [hunk 4]
```diff
@@ -710,12 +735,33 @@ func configureUpgradeClient(componentProps *HelmChartProperty, deployment *model
...
 	// This should added when we upgrade to helm ^3.13.1
-	return upgradeClient
+	return upgradeClient, nil
 }
 
-func configureUninstallClient(deployment *model.DeploymentSpec, config *action.Configuration) *action.Uninstall {
+func configureUninstallClient(ctx context.Context, componentProps *HelmChartProperty, deployment *model.DeploymentSpec, config *action.Configuration) (*action.Uninstall, error) {
 	uninstallClient := action.NewUninstall(config)
-	return uninstallClient
+	uninstallClient.Wait = componentProps.Wait
```
Comment (LeiQL) [inline api/pkg/apis/v1alpha1/providers/target/helm/helm.go:743]: Add unittest to cover the wait.

File: api/pkg/apis/v1alpha1/providers/target/helm/helm_test.go [hunk 5]
```diff
@@ -621,6 +622,95 @@ func TestHelmTargetProviderUpdateDelete(t *testing.T) {
...
+	}
+	_, err = provider.Apply(context.Background(), deployment, step, false)
+	fmt.Printf("error timeout %v", err.Error())
+	if !strings.Contains(err.Error(), "Timeout can not be negative") {
+		t.Errorf("expected error to contain 'sample', but got %s", err.Error())
+	}
+	assert.NotNil(t, err)
+}
+
+func TestHelmTargetProviderWithPositiveTimeout(t *testing.T) {
```
Comment (LeiQL) [inline api/pkg/apis/v1alpha1/providers/target/helm/helm_test.go:672]: Add a test to validate timeout error?

**Review by msftcoderdjw — APPROVED** [review]

File: api/pkg/apis/v1alpha1/providers/target/helm/helm.go (hunk 3 above)
Comment (yanjiaxin534) [reply api/pkg/apis/v1alpha1/providers/target/helm/helm.go]: OK， thanks jiawei

**Comment by yanjiaxin534** [comment]
> Of course, approved only after it builds :)
done

**Review by Haishi2016 — APPROVED** [review]

**Pull request merged**